Add files via upload

dankunis opened the pull request:
Пока оставил нерешенной одну проблему - не знаю, как получить размер файла по ссылке прежде чем загружать. (Однако нашел способ через функцию get_headers, но дело в том, что она работает только для http).

closes #4

dankunis committed:
Add files via upload

vovaksenov99 commented on Uploader.php:
```diff
@@ -0,0 +1,44 @@
+<?php
+const MaxFileSize = 5000000000000;
```
У тебя одна константа большими буквами, а другая нет

talyguryn replied:
ставь пределом 15 МБ. должно хватить для большинства задач.

n0str commented:
@dankunis It's enough to just make HTTP HEAD request, get Content-size header and disconnect.
https://stackoverflow.com/questions/2602612/php-remote-file-size-without-downloading-file

talyguryn commented on Uploader.php:
```diff
@@ -0,0 +1,44 @@
+<?php
+const MaxFileSize = 5000000000000;
+#Сейчас поставил макcимальный размер файла,который можно загрузить на Amazon - 5TB. Жду от вас нормального размера.
+const EXPANSIONS = array('image/jpg', 'image/png', 'image/jpeg', 'image/gif', 'jpg', 'png', 'jpeg', 'gif');
```
EXTENSIONS ?

talyguryn commented on Uploader.php:
```diff
@@ -0,0 +1,44 @@
+    <input type="submit" name="FileSubm" value="Accept"/>
+</form>
+<?php
+if(isset($_POST['FileSubm'])) {
```
почему ты проверяешь кнопку, а не поле с файлом?

talyguryn commented on Uploader.php:
```diff
@@ -0,0 +1,44 @@
+</form>
+<?php
+if(isset($_POST['FileSubm'])) {
+    $FileName = $_FILES['image']['name'];
```
Каждого поля может не оказаться в массиве. Надо бы добавить проверки на существование

Commit:
Все исправил

Все исправил, следуя вашим комментам, а также добавил обработку ситуации, когда человек потверждает форму, ничего не написав или не загрузив файл.

PS: Так и не нашел решения про то, как получить разммер удаленного файла.

gohabereg commented on Uploader.php:
```diff
@@ -0,0 +1,55 @@
+<?php
+const MAXFILESIZE = 125829120; #15 МБ
```
Лучше расписать как произведение 15 * 1024 * 1024 * 8. Так будет легче потом поменять значение

gohabereg commented on Uploader.php:
```diff
@@ -0,0 +1,55 @@
+<?php
+const MAXFILESIZE = 125829120; #15 МБ
+const EXTENSIONS = array('image/jpg', 'image/png', 'image/jpeg', 'image/gif', 'jpg', 'png', 'jpeg', 'gif');
```
`image/<type>` это не расширение, это MIME-type

Commit:
Сделал через классы

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,40 @@
+    private $FileExp;
+    private $FilePath;
+
+    public function Initialization()
```
Эта функция должна быть конструктором

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,40 @@
+
+class FileImage
+{
+    private $FileSize;
```
Переменные с маленькой буквы

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,40 @@
+<?php
+const MAX_FILE_SIZE = 15*1024*1024*8; #15 МБ
```
Это можно в класс занести

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    public function Checking()
```
Имена методов тоже с маленькой буквы

gohabereg replied:
Этот метод лучше сделать приватным

dankunis replied:
Если сделать приватным, то как вызывать его в Uploader.php?
И зачем это делать? (Опять же, я не спорю, а пытаюсь понять логику)

dankunis replied:
Ты хочешь, чтобы метод загрузки внутри себя вызывал метод проверки?

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,40 @@
+
+    public function Checking()
+    {
+        if(!in_array($this->FileExp, EXTENSIONS)) {
```
Разнеси все-таки проверки на разные методы `checkSize`, `checkExtension`

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,40 @@
+    public function Checking()
+    {
+        if(!in_array($this->FileExp, EXTENSIONS)) {
+            echo "Wrong file type.";
```
Лучше, наверное, кидать эксепшн

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,40 @@
+            $storage = new AWS_Storage();
+            $imgID = $storage->uploadImage($this->FilePath);
+            $imgURI = $storage->getImage($imgID);
+            echo $imgURI;
```
Лучше в методах класса не выводить ничего на страницу. Просто сделай `return $imageURL`

gohabereg commented on ImageLink.class.php:
```diff
@@ -0,0 +1,47 @@
+<?php
```
По этому классу такие же комменты как и выше

gohabereg reviewed (changes requested):
missclick

Commit:
Внес все изменения

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,60 @@
+<?php
+
+class wrongFileSize extends Exception {}
```
Мы обычно так не запариваемся, но если уж сделал класс, то назови с большой буквы

talyguryn replied:
а оно того стоит?

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,60 @@
+
+    public function upload()
+    {
+        try {
```
Эксепшены в классах обычно делаются для того, чтобы они упали снаружи. Так что try...catch тоже лучше делать снаружи, а не тут

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,60 @@
+        require_once 'lib/AWS/Storage.php';
+        $storage = new AWS_Storage();
+        $imgID = $storage->uploadImage($this->filePath);
+        $imgURI = $storage->getImage($imgID);
```
Надо будет удалить файл

gohabereg commented on FileImage.class.php:
```diff
@@ -0,0 +1,60 @@
+            $this->fileSize = $_FILES['ImageFile']['size'];
+        }
+        if(isset($_FILES['ImageFile']['tmp_name'])) {
+            $this->filePath = $_FILES['ImageFile']['tmp_name'];
```
Лучше сделать перемещение файла через `move_uploaded_file`, она проверяет, действительно ли файл был загружен на сервер через post запрос

gohabereg commented on ImageLink.class.php:
```diff
@@ -0,0 +1,57 @@
+
+    private function checkExtension()
+    {
+        if (!in_array($this->fileExp, FileImage::EXTENSIONS)) {
```
Тут ты обращаешься к другому классу, который логически никак не связан с этим. Лучше тогда подумать в сторону наследования обоих классов от одного, в котором будут методы проверки, разрешенные расширения, размер и т.д.

gohabereg commented on ImageLink.class.php:
```diff
@@ -0,0 +1,57 @@
+
+    public function __construct()
+    {
+        $this->fileName = basename($_GET['ImageLink']);
```
Лучше не брать ссылку из гет параметров, а передавать ее в конструктор как параметр

talyguryn commented on FileImage.class.php:
```diff
@@ -0,0 +1,60 @@
+class FileImage
+{
+    const MAX_FILE_SIZE = 15*1024*1024*8; #15 МБ
+    const EXTENSIONS = array('image/jpg', 'image/png', 'image/jpeg', 'image/gif', 'jpg', 'png', 'jpeg', 'gif');
```
чтобы уж совсем по красоте

```
const EXTENSIONS = array(
    'image/jpg', 'jpg',
    'image/png', 'png',
    'image/jpeg', 'jpeg',
    'image/gif', 'gif'
);

gohabereg replied:
По идее, MIME типы можно убрать

talyguryn commented on FileImage.class.php:
```diff
@@ -0,0 +1,60 @@
+        } catch (wrongFileSize $e) {
+            return $e->getMessage();
+        }
+        require_once 'lib/AWS/Storage.php';
```
у тебя же нет сейчас этих файлов. притяни мастер в свою ветку

talyguryn commented on ImageLink.class.php:
```diff
@@ -0,0 +1,57 @@
+<?php
+
+require_once 'FileImage.class.php';
```
есть ли смысл в название добавлять class?

talyguryn commented on ImageLink.class.php:
```diff
@@ -0,0 +1,57 @@
+    private $fileName;
+    private $filePath;
+
+
```
лишняя строка

talyguryn commented on ImageLink.class.php:
```diff
@@ -0,0 +1,57 @@
+
+require_once 'FileImage.class.php';
+
+class LinkImage
```
комментарий с примером использования

talyguryn commented on ImageLink.class.php:
```diff
@@ -0,0 +1,57 @@
+        $this->fileSize = $this->fileSize['Content-Length'];
+    }
+
+    private function checkExtension()
```
к каждой функции надо оставлять комментарии

talyguryn commented on ImageLink.class.php:
```diff
@@ -0,0 +1,57 @@
+        } catch (wrongFileSize $e) {
+            return $e->getMessage();
+        }
+        file_put_contents($this->fileName, file_get_contents($_GET['ImageLink']));
```
разделяй смысловые части пустыми строками

talyguryn commented on ImageLink.class.php:
```diff
@@ -0,0 +1,57 @@
+        }
+        file_put_contents($this->fileName, file_get_contents($_GET['ImageLink']));
+        $this->filePath = realpath($this->fileName);
+        require_once('lib/AWS/Storage.php');
```
может лучше подключать модули в начале файла?

talyguryn commented on Uploader.php:
```diff
@@ -0,0 +1,34 @@
+    $New_File = new LinkImage();
+    $New_File->__construct();
+    echo $New_File->upload();
+}
```
`} elseif (...) {`

talyguryn commented on Uploader.php:
```diff
@@ -0,0 +1,34 @@
+</form>
+
+<?php
+if(!empty($_GET['ImageLink']) and isset($_GET['LinkSubm'])) {
```
`if (...) {`

talyguryn commented on Uploader.php:
```diff
@@ -0,0 +1,34 @@
+
+<form method="get" action="">
+    <input type="text" name="ImageLink"/>
+    <input type="submit" name="LinkSubm" value="Accept"/>
```
возьми name проще и понятнее

talyguryn commented on Uploader.php:
```diff
@@ -0,0 +1,34 @@
+<form method="post" enctype="multipart/form-data">
+    <input type="file" name="ImageFile"/>
+    <input type="submit" name="FileSubm" value="Accept"/>
```
возьми name проще и понятнее

talyguryn reviewed (changes requested):
притяни мастер в свою ветку

Commit:
что сюда принято писать?

gohabereg commented on Image.class.php:
```diff
@@ -0,0 +1,83 @@
+    {
+        $this->fileName = basename($url['ImageLink']);
+        $this->fileExp = explode('.', $url['ImageLink']);
+        $this->fileExp = $this->fileExp[count($this->fileExp)-1];
```
Очень ненадежная проверочка будет

gohabereg commented on Image.class.php:
```diff
@@ -0,0 +1,83 @@
+class WrongFileSize extends Exception {}
+class WrongFileType extends Exception {}
+
+class Image {
```
Назови классы лучше `Uploader`, `FileUploader`, `LinkUploader`

gohabereg commented on Uploader.php:
```diff
@@ -4,14 +4,15 @@
-        $New_File->__construct();
-        echo $New_File->upload();
+require_once ('Image.class.php');
+if(is_uploaded_file($_FILES['ImageFile']['tmp_name'])) {
```
Перенеси эту проверку внутрь класса

gohabereg reviewed (changes requested):
Не жалей пустых строк между кодом

Commit:
Merge branch 'master' into capella-uploader

Commit:
притянул мастера и довел проверку расширений до ума

создал папку Uploader в lib. Все изменения там

Commit:
как идиот, случайно вставил папку из phpstorm. Прошу прощения

gohabereg commented on lib/uploader/Image.class.php:
```diff
@@ -0,0 +1,93 @@
+{
+    public function __construct($img)
+    {
+        if (is_uploaded_file($img['tmp_name'])) {
```
Перепиши в формате:
```
if (!is_uploaded_file($img['tmp_name'])) {
    throw new Exception(...);
}

Тут код, когда условие true

gohabereg commented on lib/uploader/Image.class.php:
```diff
@@ -0,0 +1,93 @@
+    {
+        $this->checkExtension();
+        $this->checkSize();
+        file_put_contents($this->fileName, file_get_contents($_GET['ImageLink']));
```
Расставь пустые строки между логическими частями

gohabereg commented on lib/uploader/Image.class.php:
```diff
@@ -0,0 +1,93 @@
+    {
+        $this->checkExtension();
+        $this->checkSize();
+        $storage = new AWS_Storage();
```
Расставь пустые строки между логическими частями

Commit:
Merge branch 'master' into capella-uploader

Commit:
Последние штрихи

gohabereg commented on lib/uploader/Image.class.php:
```diff
@@ -0,0 +1,97 @@
+    {
+        if (!is_uploaded_file($img['tmp_name'])) {
+            throw new AccessDenied("Access denied. File wasn't uploaded");
+        } else {
```
Else тут не нужен, поскольку если будет выкинут эксепшн, код дальше не будет выполнен

Commit:
убрал else

talyguryn commented on lib/uploader/Image.class.php:
```diff
@@ -0,0 +1,103 @@
+
+class AccessDenied extends Exception {}
+
+class Uploader {    #Абстрактный класс, в котором описываются допустимые расширения и максимальный размер, а так же методы проверки.
```
Комменты не так надо писать

Commit:
Added and changed some comments

Commit:
Добавил пустую строку

talyguryn committed:
First version

talyguryn committed:
Rename uploader file

talyguryn committed:
Get headers nicely

talyguryn committed:
Merge pull request #20 from codex-team/capella-uploader-update

First version
